p2p/ping.go

```diff
@@ -0,0 +1,158 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	ma "github.com/multiformats/go-multiaddr"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	pingInterval   = time.Second
+	pingProtectTag = "spacemesh-ping"
+)
+
+type pingStat struct {
+	numSuccess int
+	numFail    int
+}
+
+type Ping struct {
+	sync.Mutex
+	logger *zap.Logger
+	h      host.Host
+	peers  []peer.ID
+	pr     routing.PeerRouting
+	stats  map[peer.ID]*pingStat
+	cancel context.CancelFunc
+	eg     errgroup.Group
+}
+
+func NewPing(logger *zap.Logger, h host.Host, peers []peer.ID, pr routing.PeerRouting) *Ping {
+	p := &Ping{
+		logger: logger,
+		h:      h,
+		peers:  peers,
+		pr:     pr,
+		stats:  make(map[peer.ID]*pingStat),
+	}
+	return p
+}
+
+func (p *Ping) Start(ctx context.Context) {
```

Comment:
This context is unused.

```suggestion
func (p *Ping) Start() {
```

Reply:
I've used `context.Background()` instead of `ctx` by mistake below, will fix, thanks

Reply:
`context.Background()` is correct otherwise you have 2 ways to stop the ping: `Stop` and canceling the context passed to start 🙂

Reply:
Fixed